components/org.wso2.carbon.crypto.provider.hsm/src/main/java/org/wso2/carbon/crypto/provider/hsm/cryptoprovider/util/MechanismResolver.java

```diff
@@ -0,0 +1,306 @@
+
+        /*
+         * Sign/Verify mechanisms
+         */
+        //ECDSA sign/verify mechanisms
+        put("NONEwithECDSA", PKCS11Constants.CKM_ECDSA);
+        put("SHA1withECDSA", PKCS11Constants.CKM_ECDSA_SHA1);
+
+        //RSA sign/verify mechanisms
+        put("MD2withRSA", PKCS11Constants.CKM_MD2_RSA_PKCS);
+        put("MD5withRSA", PKCS11Constants.CKM_MD5_RSA_PKCS);
+        put("SHA1withRSA", PKCS11Constants.CKM_SHA1_RSA_PKCS);
+        put("SHA256withRSA", PKCS11Constants.CKM_SHA256_RSA_PKCS);
+        put("SHA384withRSA", PKCS11Constants.CKM_SHA384_RSA_PKCS);
+        put("SHA512withRSA", PKCS11Constants.CKM_SHA512_RSA_PKCS);
+        put("RipeMd128withRSA", PKCS11Constants.CKM_RIPEMD128_RSA_PKCS);
+        put("RipeMd160withRSA", PKCS11Constants.CKM_RIPEMD160_RSA_PKCS);
+
+        put("SHA1withRSAandMGF1", PKCS11Constants.CKM_SHA1_RSA_PKCS_PSS);
+        put("SHA256withRSAandMGF1", PKCS11Constants.CKM_SHA256_RSA_PKCS_PSS);
+        put("SHA384withRSAandMGF1", PKCS11Constants.CKM_SHA384_RSA_PKCS_PSS);
+        put("SHA512withRSAandMGF1", PKCS11Constants.CKM_SHA512_RSA_PKCS_PSS);
+
+        //DSA sign/verify mechanisms
+        put("RawDSA", PKCS11Constants.CKM_DSA);
+        put("SHA1withDSA", PKCS11Constants.CKM_DSA_SHA1);
+
+        /*
+         * Digest mechanisms
+         */
+        put("SHA1", PKCS11Constants.CKM_SHA_1);
+        put("SHA256", PKCS11Constants.CKM_SHA256);
+        put("SHA384", PKCS11Constants.CKM_SHA384);
+        put("SHA512", PKCS11Constants.CKM_SHA512);
+        put("MD2", PKCS11Constants.CKM_MD2);
+        put("MD5", PKCS11Constants.CKM_MD5);
+        put("RipeMd128", PKCS11Constants.CKM_RIPEMD128);
+        put("RipeMd160", PKCS11Constants.CKM_RIPEMD160);
+    }};
+
+    private static HashMap<Long, String> parameterRequiredMechanisms = new HashMap<Long, String>() {{
+        put(PKCS11Constants.CKM_AES_CBC, "IV16");
+        put(PKCS11Constants.CKM_AES_CBC_PAD, "IV16");
+        put(PKCS11Constants.CKM_AES_GCM, "GCM");
+
+        put(PKCS11Constants.CKM_RSA_PKCS_OAEP, "OAEP");
+
+        put(PKCS11Constants.CKM_DES3_CBC, "IV8");
+        put(PKCS11Constants.CKM_DES3_CBC_PAD, "IV8");
+
+        put(PKCS11Constants.CKM_DES_CBC, "IV8");
+        put(PKCS11Constants.CKM_DES_CBC_PAD, "IV8");
+
+        put(PKCS11Constants.CKM_BLOWFISH_CBC, "IV8");
+
+        put(PKCS11Constants.CKM_SHA1_RSA_PKCS_PSS, "PSS");
+        put(PKCS11Constants.CKM_SHA256_RSA_PKCS_PSS, "PSS");
+        put(PKCS11Constants.CKM_SHA384_RSA_PKCS_PSS, "PSS");
+        put(PKCS11Constants.CKM_SHA512_RSA_PKCS_PSS, "PSS");
+    }};
+
+    protected MechanismResolver() {
+
+    }
+
+    /**
+     * Singleton design pattern is used. Only one instance of Mechanism resolver is used for mechanism resolving.
+     *
+     * @return {@link MechanismResolver} default instance.
+     */
+    public static MechanismResolver getInstance() {
+
+        synchronized (MechanismResolver.class) {
```

Comment:
This might be a locking place when high concurrency is there. You can keep mechanismResolver initialized in global variable.